### PR TITLE
Support multiple workspaces/clients in Parcel for VSCode

### PR DESCRIPTION
**Pull request by celinanperalta**
# ↪️ Pull Request

This PR allows the Parcel for VSCode extension to report separate errors in different VSCode projects. Previously, errors in one workspace would be reported across all projects that had the extension installed.

Note that this does not include quickfix changes, which are on `lsp-ux-explorations`.

## ✔️ PR Todo

- [ ] Added/updated unit tests for this change
- [ ] Filled out test instructions (In case there aren't any unit tests)
- [ ] Included links to related issues/PRs

### CODE REVIEW

File: packages/utils/parcelforvscode/README.md [hunk 1]
```diff
@@ -6,3 +6,39 @@ This extension shows errors, warnings and other diagnostics inline in VS Code.
 
 1. Install this extension
 2. In your project, install `@parcel/reporter-lsp` and run Parcel with that reporter, e.g. `parcel src/index.html --reporter @parcel/reporter-lsp`.
+
+## Development Debugging
```
Comment (mischnic) [inline packages/utils/parcelforvscode/README.md]: This is the user-facing readme that will be visible in the extension marketplace, e.g. https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer

So this should probably be in a `CONTRIBUTING.md` or something like that

Comment (celinanperalta) [reply packages/utils/parcelforvscode/README.md]: Done

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.43s | -16.00ms |
| Cached | 268.00ms | -9.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 240.00ms | -16.00ms 🚀 |
| dist/modern/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 239.00ms | -17.00ms 🚀 |
| dist/legacy/index.ff03421b.js | 1.48kb | +0.00b | 343.00ms | -18.00ms 🚀 |
| dist/legacy/index.e9bb1616.js | 1.06kb | +0.00b | 343.00ms | -18.00ms 🚀 |
| dist/modern/index.4a29d309.js | 921.00b | +0.00b | 342.00ms | -20.00ms 🚀 |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.84s | -68.00ms |
| Cached | 401.00ms | -6.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/PermalinkedComment.e9dc4a75.js | 3.92kb | +0.00b | 323.00ms | +18.00ms ⚠️ |
| dist/UserProfile.8945a243.js | 1.38kb | +0.00b | 323.00ms | +18.00ms ⚠️ |
| dist/NotFound.8b44a81d.js | 269.00b | +0.00b | 323.00ms | +18.00ms ⚠️ |
| dist/logo.8dd07848.png | 244.00b | +0.00b | 247.00ms | +32.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/PermalinkedComment.e9dc4a75.js | 3.92kb | +0.00b | 339.00ms | +23.00ms ⚠️ |
| dist/UserProfile.8945a243.js | 1.38kb | +0.00b | 339.00ms | +23.00ms ⚠️ |
| dist/NotFound.8b44a81d.js | 269.00b | +0.00b | 339.00ms | +24.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 31.32s | -408.00ms |
| Cached | 2.13s | +13.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/media-viewer.e993d1a1.js | 536.13kb | +0.00b | 7.40s | -2.53s 🚀 |
| dist/ConfigPanelFieldsLoader.182d39bc.js | 303.43kb | +0.00b | 7.40s | +396.00ms ⚠️ |
| dist/card.d06de810.js | 138.91kb | +0.00b | 7.40s | +396.00ms ⚠️ |
| dist/ElementBrowser.e8f01080.js | 61.94kb | +0.00b | 7.40s | +396.00ms ⚠️ |
| dist/esm.271a6e34.js | 59.30kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/component-lazy.51d1dec9.js | 58.94kb | +0.00b | 4.65s | -450.00ms 🚀 |
| dist/ConfigPanelFieldsLoader.3384e28b.js | 15.74kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/ui.8c117104.js | 14.48kb | +0.00b | 7.40s | +396.00ms ⚠️ |
| dist/ConfigPanelFieldsLoader.aa7dd375.js | 13.63kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/pdfRenderer.a0700f46.js | 11.90kb | +0.00b | 7.39s | -2.43s 🚀 |
| dist/mobile-upload.3baad8e4.js | 7.79kb | +0.00b | 7.40s | +396.00ms ⚠️ |
| dist/mobile-upload.7a892a37.js | 7.79kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/ru.0cf3f40e.js | 2.81kb | +0.00b | 7.40s | +396.00ms ⚠️ |
| dist/uk.282f23b1.js | 2.76kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/codeViewerRenderer.51140ec8.js | 2.61kb | +0.00b | 7.40s | -2.63s 🚀 |
| dist/th.137e1013.js | 2.60kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/vi.b46097db.js | 2.09kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/tr.c85d90a9.js | 2.03kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/sv.1c06c95c.js | 1.98kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/zh_TW.b7c55aa6.js | 1.86kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/zh.b01fe721.js | 1.84kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/workerHasher.540c9790.js | 1.56kb | +0.00b | 7.40s | +397.00ms ⚠️ |
| dist/workerHasher.c840c607.js | 1.56kb | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/sk.4be9c93f.js | 656.00b | +0.00b | 7.40s | +396.00ms ⚠️ |
| dist/simpleHasher.c14e20b4.js | 589.00b | +0.00b | 7.40s | +397.00ms ⚠️ |
| dist/simpleHasher.23db7a52.js | 589.00b | +0.00b | 7.40s | +395.00ms ⚠️ |
| dist/ro.8d5b380a.js | 482.00b | +0.00b | 5.29s | -1.60s 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.88s | +32.00ms |
| Cached | 306.00ms | -9.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/379395695165047365)**

File: packages/reporters/lsp-reporter/src/LspReporter.js [hunk 2]
```diff
@@ -67,58 +68,109 @@ let bundleGraphDeferrable =
...
+let lspStarted = false;
+let watchStartPromise;
+
+async function watchLspActive(): Promise<FSWatcher> {
+  const lspFileName = 'lsp-server';
+
+  // Check for lsp-server when reporter is first started
+  try {
+    await fs.promises.access(
+      path.join(BASEDIR, 'lsp-server'),
```
Comment (lettertwo) [inline packages/reporters/lsp-reporter/src/LspReporter.js]:
```suggestion
      LSP_SENTINEL_FILE,
```

File: packages/reporters/lsp-reporter/src/LspReporter.js [hunk 3]
```diff
@@ -67,58 +68,109 @@ let bundleGraphDeferrable =
...
+  } catch {
+    //
+  }
+
+  return fs.watch(BASEDIR, (eventType: string, filename: string) => {
+    switch (eventType) {
+      case 'rename':
+        if (filename === lspFileName) {
+          fs.access(
+            path.join(BASEDIR, 'lsp-server'),
```
Comment (lettertwo) [inline packages/reporters/lsp-reporter/src/LspReporter.js]:
```suggestion
            LSP_SENTINEL_FILE,
```

File: packages/utils/parcel-lsp/src/LspServer.ts [hunk 4]
```diff
@@ -263,27 +273,35 @@ function createClient(metafilepath: string) {
...
   let filepath = path.join(BASEDIR, filename);
-  createClient(filepath);
-  // console.log('connected initial', filepath);
+  const contents = parseMetafile(filepath);
+  const {projectRoot} = contents;
+
+  if (WORKSPACE_ROOT === projectRoot) {
+    createClient(filepath, contents);
+  }
+  console.log('connected initial', filepath);
```
Comment (lettertwo) [inline packages/utils/parcel-lsp/src/LspServer.ts]:
```suggestion
```

File: packages/utils/parcel-lsp/src/LspServer.ts [hunk 5]
```diff
@@ -295,15 +313,21 @@ watcher.subscribe(BASEDIR, async (err, events) => {
...
+      console.log('connected watched', event.path);
     } else if (event.type === 'delete' && event.path.endsWith('.json')) {
       let existing = clients.get(event.path);
-      // console.log('existing', event.path, existing);
+      console.log('existing', event.path, existing);
       if (existing) {
         clients.delete(event.path);
         existing.connection.end();
-        // console.log('disconnected watched', event.path);
+        console.log('disconnected watched', event.path);
```
Comment (lettertwo) [inline packages/utils/parcel-lsp/src/LspServer.ts]:
```suggestion
```

File: packages/utils/parcel-lsp/src/LspServer.ts [hunk 6]
```diff
@@ -295,15 +313,21 @@ watcher.subscribe(BASEDIR, async (err, events) => {
...
-      createClient(event.path);
-      // console.log('connected watched', event.path);
+      const file = fs.readFileSync(event.path, 'utf-8');
+      const contents = parseMetafile(file);
+      const {projectRoot} = contents;
+
+      if (WORKSPACE_ROOT === projectRoot) {
+        createClient(event.path, contents);
+      }
+      console.log('connected watched', event.path);
```
Comment (lettertwo) [inline packages/utils/parcel-lsp/src/LspServer.ts]:
```suggestion
```

File: packages/utils/parcel-lsp/src/LspServer.ts [hunk 7]
```diff
@@ -295,15 +313,21 @@ watcher.subscribe(BASEDIR, async (err, events) => {
...
     } else if (event.type === 'delete' && event.path.endsWith('.json')) {
       let existing = clients.get(event.path);
-      // console.log('existing', event.path, existing);
+      console.log('existing', event.path, existing);
```
Comment (lettertwo) [inline packages/utils/parcel-lsp/src/LspServer.ts:324]:
```suggestion
```

File: packages/utils/parcelforvscode/vscode-extension-TODO.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
+Packages:
+
+- [@parcel/reporter-lsp](./packages/reporters/lsp-reporter/)
+- [parcel-for-vscode](./packages/utils/parcelforvscode/)
+- [@parcel/lsp](./packages/utils/parcel-lsp/)
+- [@parcel/lsp-protocol](./packages/utils/parcel-lsp-protocol)
```
Comment (lettertwo) [inline packages/utils/parcelforvscode/vscode-extension-TODO.md:6]:
```suggestion
- [@parcel/reporter-lsp](../../reporters/lsp-reporter/)
- [parcel-for-vscode](../parcelforvscode/)
- [@parcel/lsp](../parcel-lsp/)
- [@parcel/lsp-protocol](../parcel-lsp-protocol)
```

File: packages/reporters/lsp-reporter/src/LspReporter.js [hunk 9]
```diff
@@ -67,58 +68,109 @@ let bundleGraphDeferrable =
...
-          let pid = parseInt(filename.slice('parcel-'.length), 10);
-          let resultList = await lookupPid({pid});
-          if (resultList.length > 0) continue;
-          fs.unlinkSync(path.join(BASEDIR, filename));
-          ignoreFail(() =>
-            fs.unlinkSync(path.join(BASEDIR, filename + '.json')),
+let watchStarted = false;
+let lspStarted = false;
+let watchStartPromise;
+
```
Comment (lettertwo) [inline packages/reporters/lsp-reporter/src/LspReporter.js:74]:
```suggestion

const LSP_SENTINEL_FILENAME = 'lsp-server';
const LSP_SENTINEL_FILE = path.join(BASEDIR, LSP_SENTINEL_FILENAME);

```

File: packages/reporters/lsp-reporter/src/LspReporter.js [hunk 10]
```diff
@@ -67,58 +68,109 @@ let bundleGraphDeferrable =
...
+async function watchLspActive(): Promise<FSWatcher> {
+  const lspFileName = 'lsp-server';
```
Comment (lettertwo) [inline packages/reporters/lsp-reporter/src/LspReporter.js]:
```suggestion
```

File: packages/reporters/lsp-reporter/src/LspReporter.js [hunk 11]
```diff
@@ -67,58 +68,109 @@ let bundleGraphDeferrable =
...
+
```
Comment (lettertwo) [inline packages/reporters/lsp-reporter/src/LspReporter.js:77]:
```suggestion
```

File: packages/reporters/lsp-reporter/src/LspReporter.js [hunk 12]
```diff
@@ -67,58 +68,109 @@ let bundleGraphDeferrable =
...
+    );
+    lspStarted = true;
+  } catch {
+    //
+  }
+
+  return fs.watch(BASEDIR, (eventType: string, filename: string) => {
+    switch (eventType) {
+      case 'rename':
+        if (filename === lspFileName) {
```
Comment (lettertwo) [inline packages/reporters/lsp-reporter/src/LspReporter.js]:
```suggestion
        if (filename === LSP_SENTINEL_FILENAME) {
```

File: packages/utils/parcel-lsp/src/LspServer.ts [hunk 13]
```diff
@@ -36,7 +36,14 @@ import {
...
 } from '@parcel/lsp-protocol';
 
+type Metafile = {
+  projectRoot: string;
+  pid: typeof process['pid'];
+  argv: typeof process['argv'];
+};
+
 const connection = createConnection(ProposedFeatures.all);
+const WORKSPACE_ROOT = process.cwd();
```
Comment (lettertwo) [inline packages/utils/parcel-lsp/src/LspServer.ts]:
```suggestion
const LSP_SENTINEL_FILENAME = 'lsp-server';
```

File: packages/utils/parcel-lsp/src/LspServer.ts [hunk 14]
```diff
@@ -263,27 +273,35 @@ function createClient(metafilepath: string) {
...
   sendDiagnosticsRefresh();
-  clients.set(metafile, result);
+  clients.set(JSON.stringify(metafile), result);
 }
 
 // Take realpath because to have consistent cache keys on macOS (/var -> /private/var)
 const BASEDIR = path.join(fs.realpathSync(os.tmpdir()), 'parcel-lsp');
 fs.mkdirSync(BASEDIR, {recursive: true});
+
+fs.writeFileSync(path.join(BASEDIR, `lsp-server`), 'help');
```
Comment (lettertwo) [inline packages/utils/parcel-lsp/src/LspServer.ts]:
```suggestion
fs.writeFileSync(path.join(BASEDIR, LSP_SENTINEL_FILENAME), '');
```

File: packages/utils/parcelforvscode/vscode-extension-TODO.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+- [x] need to not wait for connections
+- [x] language server shuts down and kills our process when the extension is closed
+- [x] handle the case where parcel is started after the extension is running
+- [x] handle the case where extension is started while parcel is running
+- [x] support multiple parcels
+- [x] show prior diagnostics on connection
+- [x] only connect to parcels that match the workspace
+- [ ] show parcel diagnostic hints
+- [ ] implement quick fixes (requires Parcel changes?)
+- [x] cleanup LSP server sentinel when server shuts down
```
Comment (lettertwo) [inline packages/utils/parcelforvscode/vscode-extension-TODO.md:19]: Is this done? How does the sentinel file get cleaned up?

**Review by lettertwo — CHANGES_REQUESTED** [review]

File: packages/utils/parcelforvscode/vscode-extension-TODO.md (hunk 15 above)
Comment (celinanperalta) [reply packages/utils/parcelforvscode/vscode-extension-TODO.md:19]: It's cleaned up by `client.onClose` in `LspServer.js`

Comment (lettertwo) [reply packages/utils/parcelforvscode/vscode-extension-TODO.md:19]: 🤔 Maybe i'm missing something, but I don't see anywhere where the sentinel file is unlinked. AFAICT, the sentinel file will just exist in the tmp dir forever once the lsp server has been started once.

If that's the case, i think it's fine for now. We can just uncheck this task to remind ourselves we still need to do it.

We might want to enforce a singleton condition on the lsp server, as well, so the sentinel might have change a bit, anyway.

Comment (celinanperalta) [reply packages/utils/parcelforvscode/vscode-extension-TODO.md:19]: Added [this](https://github.com/parcel-bundler/parcel/pull/9278/files#diff-f3c6654ce2aa9ac5aa019e5af5506352169f4453e99e57ef2dba06aeab2f81ccR55) to the `deactivate` function in `extension.ts`. I think we should only remove the sentinel file when the extension is deactivated.

**Review by lettertwo — APPROVED** [review]

**Pull request merged**